File: 2341-maximum-number-of-pairs-in-array/2341-maximum-number-of-pairs-in-array.py
class Solution:
    def numberOfPairs(self, nums: List[int]) -> List[int]:
        hash_map = {}
        for num in nums:
            if num in hash_map:
                hash_map[num] += 1
            else:
                hash_map[num] = 1
        
        pairs, leftover = 0, 0
        
        for key in hash_map:
            pairs += hash_map[key] // 2
            leftover += hash_map[key] % 2
        
        return [pairs,leftover]

File: 2341-maximum-number-of-pairs-in-array/2341-maximum-number-of-pairs-in-array.java
class Solution {
    public int[] numberOfPairs(int[] nums) {
        HashMap<Integer,Integer> hashMap = new HashMap<>();
        for (int num : nums) {
            Integer count = hashMap.getOrDefault(num, 0);
            hashMap.put(num, count + 1);
        }

        int pairCount = 0 ;
        int leftover = 0;
        for (HashMap.Entry<Integer,Integer> entry: hashMap.entrySet()) {
            Integer value = entry.getValue();
            pairCount += value / 2;
            leftover += value % 2;
        }

        return new int[]{pairCount, leftover};
    }
}